### ISSUE TITLE
iOS (Unity support) marshaling delegates not supported for IL2Cpp

### ISSUE THREAD

**Issue by PawelWitak** [grpc/grpc#17587]
Currently I am trying to use Google Text-to-Speech with Nuget packages which are using for gRPC. 
I see the experimental support for iOS, but when I try to use that library I am encountering "IL2CPP doesn't support marshalling delegates". How to resolve that issue?

**Comment by PawelWitak** [comment]
Can anyone help me?

**Comment by PawelWitak** [comment]
I have managed to find the same issue on Unity forum, so maybe this [link](https://forum.unity.com/threads/grpc-call-interceptors-are-not-supported-on-il2cpp.585583) will allow someone to understand the problem.

**Comment by PawelWitak** [comment]
Yep, it cannot be just static as static method cannot access instance fields. There is need to refactor more code, not only that one function. Hopefully @jsmouret your answer would be impuls for author to fix it.

**Comment by ryangreen** [comment]
I am running into the same problem when using CloudSpeech.v1 from nuget with the grpc unity library.  Works on android and desktop but fails on iOS since iOS forces 64bit. Could really use some guidance @ncteisen  as supporting iOS essential to our work.

**Comment by ryangreen** [comment]
I then had to rebuild Grpc.Core from source for the csharp branch and copy the result into the plugins/Android folder

**Comment by ryangreen** [comment]
I also had to replace Newtonsoft.Json with code built to support IL2cpp backends... there's an asset available here: https://assetstore.unity.com/packages/tools/input-management/json-net-for-unity-11347

**Comment by ryangreen** [comment]
and then I added all the dll's i'm including in the ios build to a link.xml file in my plugins folder...

```
<linker>
  <assembly fullname="Google.Apis.Auth" preserve="all"/>
  <assembly fullname="Google.Api.CommonProtos" preserve="all"/>
  <assembly fullname="Google.Api.Gax" preserve="all"/>
  <assembly fullname="Google.Api.Gax.Grpc" preserve="all"/>
  <assembly fullname="Google.Apis" preserve="all"/>
  <assembly fullname="Google.Apis.Auth.PlatformServices" preserve="all"/>
  <assembly fullname="Google.Apis.Core" preserve="all"/>
  <assembly fullname="Google.Apis.Dialogflow.v2" preserve="all"/>
  <assembly fullname="Google.Apis.PlatformServices" preserve="all"/>
  <assembly fullname="Google.Cloud.Speech.V1" preserve="all"/>
  <assembly fullname="Google.LongRunning" preserve="all"/>
  <assembly fullname="Google.Protobuf" preserve="all"/>
  <assembly fullname="Grpc.Auth" preserve="all"/>
  <assembly fullname="Grpc.Core" preserve="all"/>
  <assembly fullname="libgrpc_csharp_ext" preserve="all"/>
  <assembly fullname="Microsoft.Win32.Primitives" preserve="all"/>
  <assembly fullname="Newtonsoft.Json" preserve="all"/>
  <assembly fullname="System.Diagnostics.DiagnosticSource" preserve="all"/>
  <assembly fullname="System.Interactive.Async" preserve="all"/>
  <assembly fullname="System.IO" preserve="all"/>
  <assembly fullname="System.Runtime" preserve="all"/>
  <assembly fullname="System.Security.Cryptography.Algorithms" preserve="all"/>
  <assembly fullname="System.Security.Cryptography.Encoding" preserve="all"/>
  <assembly fullname="System.Security.Cryptography.Primitives" preserve="all"/>
  <assembly fullname="System.Security.Cryptography.X509Certificates" preserve="all"/>
  <assembly fullname="System">
    <type fullname="System.ComponentModel.TypeConverter" preserve="all" />
    <!-- <namespace fullname="System.ComponentModel" preserve="all" /> -->
  </assembly>
</linker>

**Comment by PawelWitak** [comment]
I have checked that above fix works like a charm (so far I checked only on iOS simulator but I suppose on the real device also should work). So in my opinion this is additional prove to put the fix in main repository.

**Comment by jtattermusch** [comment]
@ryangreen  @PawelWitak   feel free to open a pull request with the necessary patches. The potential problem is we don't want to resort to dirty hacks that would lower the quality of the C# code just because Unity happens to be very special runtime from many respects. But if we can find a clean way to work around the problem, we will be happy to accept the patch.

**Comment by PawelWitak** [comment]
I am not sure this is dirty hack or workaround as I have checked that this patch works great even on real device (not only on simulator). So @ryangreen as it is your patch, maybe you will open a pull request?

**Comment by jtattermusch** [comment]
I looked at @ryangreen's snippet more closely:
The idea of using [MonoPInvokeCallback(typeof(NativeMetadataInterceptor))] and a static method is the correct approach, but the solution is incomplete. Using `static NativeMetadataCredentialsPlugin instance` clearly doesn't allow more than one instance of NativeMetadataCredentialsPlugin to be used at a time and that's broken. So the snippet is useful as work-in-progress proof of concept, but it's definitely not acceptable into upstream as is.

**Comment by jtattermusch** [comment]
I think this was fixed by #18327. The fix will be in the upcoming release v1.20.0 (or you can test using nightly build packages from packages.grpc.io, they should be available soon).
Can you confirm that this was fixed?

**Comment by jsmouret** [comment]
Confirmed, thanks :)
